IPv4/single: connect() should return immediately after a protocol error and other things

htibosch opened the pull request:
Description
-----------
A TCP connection can be established in either two ways: accept() and connect().

When using connect(), the socket is created and owned by the application.

When using accept(), the new socket is created and temporarily owned by the IP-task, and in the comments it is called an "orphaned socket". Once the connection is established, the function accept() will pass the ownership to the application.

The stack has a mechanism that checks sockets for inactivity: `ipconfigTCP_HANG_PROTECTION`. After `ipconfigTCP_HANG_PROTECTION_TIME` seconds of inactivity an orphaned socket will be closed.

There were a few reasons to create this PR:

1) When a protocol error occurs during a `FreeRTOS_connect()`, the API should stop waiting and return `-pdFREERTOS_ERRNO_ENOTCONN` to the application. At this moment it will only return when the timeout has been reached.

2) When a protocol error occurs during a passive connection, the socket will hang until the "hang protection" will close the socket. When a fatal protocol error occurs the client socket should be closed immediately.

3) The re-use option `FREERTOS_SO_REUSE_LISTEN_SOCKET`: when the listening socket is configured to turn itself into a "client socket", it may not be closed after a protocol error or after "hang protection". The reusable socket will be put into the listening state again and it will be able to receive a new connection.


Many thanks to [Am Paschal](https://forums.freertos.org/u/AmPaschal), who wrote the post "FreeRTOS connect and accept functions (in FreeRTOS+TCP) do not return even after the TCP connection closes" on the [FreeRTOS forum](https://forums.freertos.org/t/freertos-connect-and-accept-functions-in-freertos-tcp-do-not-return-even-after-the-tcp-connection-closes).

Also thanks to [Alex "oberondoro"](https://forums.freertos.org/u/oberondoro) who wrote the post "Sockets end in state eCLOSE_WAIT" on the [FreeRTOS forum](https://forums.freertos.org/t/sockets-end-in-state-eclose-wait).

And thanks to [Hubert Sack](https://forums.freertos.org/u/hs4FreeRtos) who explained the problem which occurs when a reusable socket has a protocol error or hang in the SYN phase.

Test Steps
-----------

Testing this PR was a bit complicated because of the many possibilities:

1) Test an outgoing connection using connect()
2) Test an incoming connection with a normal listening socket
3) Test an incoming connection with a reusable socket

X. Induce a protocol error during the SYN phase
Y. Induce a time during the SYN phase and get a hang protection.

I did 6 tests in total: 1X, 2X, 3X, and 1Y, 2Y, and 3Y.

When the listening socket is reusable, it should return to its listening state and be able to receive a new connection.
Normal client sockets should be closed when a protocol error or a timeout occurs so that the resources become available again.
`FreeRTOS_connect()` should return immediately when a protocol error occurs, or when the socket stays inactive for a long time.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

AniruddhaKanhere commented on source/FreeRTOS_Sockets.c:
```diff
@@ -3139,7 +3139,18 @@ void vSocketWakeUpUser( FreeRTOS_Socket_t * pxSocket )
 
                 /* Go sleeping until we get any down-stream event */
-                ( void ) xEventGroupWaitBits( pxSocket->xEventGroup, ( EventBits_t ) eSOCKET_CONNECT, pdTRUE /*xClearOnExit*/, pdFALSE /*xWaitAllBits*/, xRemainingTime );
+                EventBits_t uxEvents = xEventGroupWaitBits( pxSocket->xEventGroup,
```
Maybe declare the variable at top of the block? Maybe just after `for( ; ; )`?

htibosch replied:
Done so, thanks

AniruddhaKanhere commented on source/FreeRTOS_Sockets.c:
```diff
@@ -3287,7 +3298,11 @@ void vSocketWakeUpUser( FreeRTOS_Socket_t * pxSocket )
                 }
 
                 /* Go sleeping until we get any down-stream event */
```
```suggestion
                /* Put the calling task to 'sleep' until a down-stream event is received. */
```

AniruddhaKanhere commented on source/FreeRTOS_TCP_IP.c:
```diff
@@ -272,39 +272,56 @@
+            FreeRTOS_printf( ( "Move from %s to %s\n",
+                               FreeRTOS_GetTCPStateName( xPreviousState ),
+                               FreeRTOS_GetTCPStateName( eTCPState ) ) );
+            bBefore = pdTRUE;
```
```suggestion
            /* Set the flag to show that it was connected before and that the
             * status has changed now. This will cause the control flow to go 
             * in the below if condition.*/
            bBefore = pdTRUE;
```

AniruddhaKanhere commented on source/FreeRTOS_TCP_IP.c:
```diff
@@ -393,30 +413,48 @@
+                case eSYN_RECEIVED: /* 4 (server) waiting for a confirming connection request */
+                    FreeRTOS_printf( ( "Restoring a reuse socket port %u\n", pxSocket->usLocalPort ) );
+
+                    /* Go back into list mode. Set the TCP status to 'eCLOSED',
```
```suggestion
                    /* Go back into listening mode. Set the TCP status to 'eCLOSED',
```

AniruddhaKanhere commented on source/FreeRTOS_TCP_IP.c:
```diff
@@ -393,30 +413,48 @@
+                    /* Go back into list mode. Set the TCP status to 'eCLOSED',
+                     * otherwise FreeRTOS_listen() will refuse the action. */
+                    pxSocket->u.xTCP.eTCPState = eCLOSED;
+                    ( void ) FreeRTOS_listen( ( Socket_t ) pxSocket, pxSocket->u.xTCP.usBacklog );
```
MISRA will not like this.
`FreeRTOS_listen` -> `vTCPStateChange` -> `FreeRTOS_listen`.
How is this recursion stopped? What is to say that it won't keep calling itself?

htibosch replied:
I was aware of it, but:

- When `vTCPStateChange()` calls `FreeRTOS_listen()`, the current TCP state is either `eSYN_FIRST` or `eSYN_RECEIVED`.
- `FreeRTOS_listen()` will call `vTCPStateChange()` while the current TCP state is `eCLOSED`. That can not lead to a new call to `listen()`.

Of course I could copy some of the listen() code into `vTCPStateChange()`, but I prefer to call the original API.

AniruddhaKanhere commented:
/bot run uncrustify

htibosch commented:
About "Avoid a recursive call to vTCPStateChange()" :

Earlier we introduced `vSocketCloseNextTime()` so that a call to `closesocket()` could be postponed.

Now in order to avoid a recursive call to `vTCPStateChange()`, I introduced `vSocketListenNextTime()`.

Just before the IP-task starts waiting for the next message, `FreeRTOS_listen()` may be called.

This may happen when a reusable listening socket gets a timeout or a protocol error before it is connected.
~~~c
static FreeRTOS_Socket_t * xSocketToListen = NULL;
void vSocketListenNextTime( FreeRTOS_Socket_t * pxSocket )
{
    if( ( xSocketToListen != NULL ) && ( xSocketToListen != pxSocket ) )
    {
        ( void ) FreeRTOS_listen( ( Socket_t ) xSocketToListen, xSocketToListen->u.xTCP.usBacklog );
    }

    xSocketToListen = pxSocket;
}
~~~
I tested the changes in all possible combinations:

1. By calling connect()
2. By calling accept() with a reusable socket
3. By calling accept() with a normal socket

There were 3 scenarios:

1. The connection went well
2. A protocol error occurred
3. A timeout occurred

The results were as expected, in situations 2.2 and 2.3, a delayed call to `FreeRTOS_listen()` was made, and the listening socket got restored.

AniruddhaKanhere commented:
/bot run uncrustify

AniruddhaKanhere commented:
/bot run uncrustify

cobusve commented on source/FreeRTOS_TCP_IP.c:
```diff
@@ -272,39 +299,60 @@
+        {
+            /* A socket was in the connecting phase but something
+             * went wrong and it should be closed. */
+            FreeRTOS_printf( ( "Move from %s to %s\n",
```
It seems to me that this should probably be a debug_printf ?

cobusve commented on source/FreeRTOS_Sockets.c:
```diff
@@ -3139,7 +3141,18 @@ void vSocketWakeUpUser( FreeRTOS_Socket_t * pxSocket )
+                if( ( uxEvents & eSOCKET_CLOSED ) != 0U )
+                {
+                    xResult = -pdFREERTOS_ERRNO_ENOTCONN;
+                    FreeRTOS_printf( ( "FreeRTOS_connect() stopped due to an error\n" ) );
```
I think this is another one?

AniruddhaKanhere replied:
done

cobusve commented on source/FreeRTOS_TCP_IP.c:
```diff
@@ -393,30 +444,51 @@
+            {
+                case eSYN_FIRST:    /* 3 (server) Just created, must ACK the SYN request */
+                case eSYN_RECEIVED: /* 4 (server) waiting for a confirming connection request */
+                    FreeRTOS_printf( ( "Restoring a reuse socket port %u\n", pxSocket->usLocalPort ) );
```
Is this another debug message?

AniruddhaKanhere replied:
done
